kong/cluster_events/strategies/postgres.lua

```diff
@@ -113,23 +126,35 @@ end
 function _M:select_interval(channels, min_at, max_at)
   local n_chans = #channels
   local p_chans = new_tab(n_chans, 0)
-  local p_minat = self.connector:escape_literal(tonumber(fmt("%.3f", min_at)))
-  local p_maxat = self.connector:escape_literal(tonumber(fmt("%.3f", max_at)))
 
   for i = 1, n_chans do
     p_chans[i] = self.connector:escape_literal(channels[i])
   end
 
-  local query_template = fmt(SELECT_INTERVAL_QUERY,
-                             concat(p_chans, ", "),
-                             p_minat,
-                             p_maxat,
-                             self.page_size,
-                             "%s")
+  p_chans = concat(p_chans, ", ")
 
-  local page = 0
+  local p_minat = self.connector:escape_literal(tonumber(fmt("%.3f", min_at or 0)))
+  local p_maxat
+  if max_at then
+    p_maxat = fmt("TO_TIMESTAMP(%s) AT TIME ZONE 'UTC'",
+                  self.connector:escape_literal(tonumber(fmt("%.3f", max_at))))
+  else
+    p_maxat = "CURRENT_TIMESTAMP(3) AT TIME ZONE 'UTC'"
+  end
 
+  local query_template = fmt(SELECT_INTERVAL_QUERY, p_chans,
+                                                    p_minat,
+                                                    p_maxat,
+                                                    self.page_size,
+                                                    "%s")
+
+  local page = 0
+  local last_page
   return function()
+    if last_page then
```

Comment:
I think the `last_page` optimization should not be included.

Reasoning: people can "take their time" while iterating over the `select_interval` function. By that I mean they could read the first page, keep the token, wait 3 minutes, read the second page, keep the token, wait 3 minutes, then read the next page, etc.

Even if a page returned only 1 record on the last read, in those 3 minutes a lot of things can have happened and I should get those events.

I'm using "3 minutes" as a random "infinitely long" period, but if the gods of concurrency and CPUs align themselves, even a delay of 1ms (perhaps with some IO in another processor sprinkled in) can spell the difference.

Reply:
Discussed this privately and was convinced that this is not an issue on this particular case :we are the only users of this API, we consume it immediately, and even if we weren't the early `nil` just means that some events might be queued and processed on the next call to "pull events".